pop_tools/regrid.py

```diff
@@ -0,0 +1,432 @@
+    out_lons[1:, 1:] = corn_lon[:, :, 0]
+    out_lats[1:, 1:] = corn_lat[:, :, 0]
+
+    # Assign the northwest corner information
+    out_lons[1:, :-1] = corn_lon[:, :, 1]
+    out_lats[1:, :-1] = corn_lat[:, :, 1]
+
+    # Assign the southwest corner information
+    out_lons[:-1, :-1] = corn_lon[:, :, 2]
+    out_lats[:-1, :-1] = corn_lat[:, :, 2]
+
+    # Assign the southeast corner information
+    out_lons[:-1, 1:] = corn_lon[:, :, 3]
+    out_lats[:-1, 1:] = corn_lat[:, :, 3]
+
+    return out_lats, out_lons
+
+
+def _add_region_mask(ds, grid, cell_center='TLAT'):
+    """
+    Grabs and combines a grid
+    """
+
+    # Grab the associated grid
+    if not grid:
+        raise TypeError('User did not input a grid to use')
+    else:
+        grid = get_grid(grid)
+
+    return xr.merge([ds, grid[['REGION_MASK', 'region_name']]])
+
+
+def _get_default_filename(src_grid, dst_grid, method):
+
+    # Get the source grid shape
+    src_shape = src_grid.lat.shape
+
+    # Get the destination grid shape
+    dst_shape = dst_grid.lat.shape
+
+    filename = f'{method}_{src_shape[0]}x{src_shape[1]}_{dst_shape[0]}x{dst_shape[1]}.nc'
+
+    return filename
+
+
+def _generate_weights(src_grid, dst_grid, method, weight_file=None):
+    """
+    Generate regridding weights by calling xESMF
+    """
+
+    # Allow user to input weights file, if there is not one, use default check
+    if weight_file is None:
+        weight_file = _get_default_filename(src_grid, dst_grid, method)
+
+    # Check to see if the weights file already exists - if not, generate weights
+    if not os.path.exists(weight_file):
+        regridder = xe.Regridder(src_grid, dst_grid, method, ignore_degenerate=True)
+        print(f'Saving weights file: {os.path.abspath(weight_file)}')
+        regridder.to_netcdf(weight_file)
+
+    else:
+        regridder = xe.Regridder(
+            src_grid, dst_grid, method, weights=weight_file, ignore_degenerate=True
+        )
+
+    return regridder
+
+
+# Setup method for regridding a dataarray
+def _regrid_dataset(da_in, dst_grid, regrid_method=None):
+    src_grid = _convert_to_xesmf(da_in)
+
+    # If the user does not specify a regridding method, use default conservative
+    if regrid_method is None:
+        regridder = _generate_weights(src_grid, dst_grid, 'conservative')
+
+    else:
+        regridder = _generate_weights(src_grid, dst_grid, regrid_method)
+
+    # Regrid the input data array, assigning the original attributes
+    da_out = regridder(src_grid)
+    da_out.attrs = da_in.attrs
+
+    return da_out.drop(['lat_b', 'lon_b'])
+
+
+def _convert_to_xesmf(ds):
+
+    if isinstance(ds, xr.DataArray):
+        grid_center_lat = ds.cf['latitude'].name
+        grid_center_lon = ds.cf['longitude'].name
+        ds = ds.to_dataset()
+
+    elif isinstance(ds, xr.Dataset):
+        var = list(ds.variables)[0]
+        grid_center_lat = ds[var].cf['latitude'].name
+        grid_center_lon = ds[var].cf['longitude'].name
```

Comment:
We are using `cf_xarray` accessor here, but I don't see `cf_xarray` in the imports. Is this working as expected without the `cf_xarray` import?

Reply:
also `list(ds.variables)[0]`?

I think the only sane way to do this is to look at loop over DataArrays and check `"TLAT" in DataArray.attrs["coordinates"]`.So take an input dataset then split it in to two: those with coordinates TLAT, TLONG; and then those with ULAT, ULONG; process the two datasets separately then combine them to a single dataset and return that.

Reply:
@dcherian This logic is handled within the `to_uniform_grid` function...

Reply:
@mgrover1, as Deepak pointed out, this line is going to result in errors when the "first" grabbed variable doesn't have spatial dimensions...

> var = list(ds.variables)[0]

Could we make this more robust?

Reply:
I can go ahead and remove the check to see if it is a dataset since it *should* be a data array, and this is a function that is meant to used internally... in to_uniform_grid, we check to see if nlat, nlon, and TLAT are in the dimensions, creating a list of variables. These variables are then passed into prep_for_xesmf function, iterating over those variables.